File: starboard/android/apk/app/src/main/java/dev/cobalt/media/VideoDecoderCache.java
// Copyright 2021 The Cobalt Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package dev.cobalt.media;

import android.media.MediaCodecInfo;
import android.media.MediaCodecInfo.CodecCapabilities;
import android.media.MediaCodecInfo.VideoCapabilities;
import android.media.MediaCodecList;
import android.util.Range;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Cache manager for maintaining a video decoder cache. Threadsafe. */
public class VideoDecoderCache {
  /**
   * Cache entry in VideoDecoderCache. Cached expensive API calls that are repeatedly made while the
   * H5 player determines codec support.
   */
  public static class CachedDecoder {
    CachedDecoder(MediaCodecInfo info, CodecCapabilities codecCapabilities, String mimeType) {
      this.info = info;
      this.mimeType = mimeType;
      this.codecCapabilities = codecCapabilities;
      this.videoCapabilities = codecCapabilities.getVideoCapabilities();
      this.supportedWidths = videoCapabilities.getSupportedWidths();
      this.supportedHeights = videoCapabilities.getSupportedHeights();
      this.bitrateRange = videoCapabilities.getBitrateRange();
    }

    public MediaCodecInfo info;
    public String mimeType;
    public CodecCapabilities codecCapabilities;
    public VideoCapabilities videoCapabilities;
    public Range<Integer> supportedWidths;
    public Range<Integer> supportedHeights;
    public Range<Integer> bitrateRange;
  }

  private static final int DEFAULT_CACHE_TTL_MILLIS = 1000;

  private static Map<String, List<CachedDecoder>> cache = new HashMap<>();
  private static long lastCacheUpdateAt = 0;

  private static synchronized boolean isExpired(int cacheTtlOverride) {
    long cacheTtl = cacheTtlOverride >= 0 ? cacheTtlOverride : DEFAULT_CACHE_TTL_MILLIS;
    return System.currentTimeMillis() - lastCacheUpdateAt >= cacheTtl;
  }

  private static void refreshDecoders() {
    cache.clear();
    // Note: MediaCodecList is sorted by the framework such that the best decoders come first.
    // This order is maintained in the cache.
    for (MediaCodecInfo codecInfo : new MediaCodecList(MediaCodecList.ALL_CODECS).getCodecInfos()) {
      // Don't cache encoders.
      if (codecInfo.isEncoder()) {
        continue;
      }
      for (String mimeType : codecInfo.getSupportedTypes()) {
        CodecCapabilities codecCapabilities = codecInfo.getCapabilitiesForType(mimeType);
        // Don't cache decoders that can't do video.
        if (codecCapabilities.getVideoCapabilities() == null) {
          continue;
        }
        List<CachedDecoder> decoders = cache.get(mimeType);
        if (decoders == null) {
          decoders = new ArrayList<>();
          cache.put(mimeType, decoders);
        }
        decoders.add(new CachedDecoder(codecInfo, codecCapabilities, mimeType));
      }
    }
    lastCacheUpdateAt = System.currentTimeMillis();
  }

  /**
   * Returns a list of decoders that are cabable of decoding the media denoted by {@code mimeType}.
   *
   * @param mimeType The mime type the returned decoders should be able to decode.
   * @param cacheTtlOverride Overrides the default cache TTL if passed a non-negative number. 0
   *     disables the cache entirely.
   * @return A list of decoders.
   */
  static List<CachedDecoder> getCachedDecoders(String mimeType, int cacheTtlOverride) {
    synchronized (cache) {
      if (isExpired(cacheTtlOverride)) {
        refreshDecoders();
      }
      return cache.containsKey(mimeType)
          ? cache.get(mimeType)
          : Collections.<CachedDecoder>emptyList();
    }
  }
}

File: starboard/android/shared/video_decoder_cache.cc
// Copyright 2026 The Cobalt Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "starboard/android/shared/video_decoder_cache.h"

#include "base/no_destructor.h"
#include "starboard/common/log.h"
#include "starboard/common/media.h"
#include "starboard/common/player.h"

namespace starboard::android::shared {

namespace {
constexpr int kMaxCacheSize = 2;
}  // namespace

// static
VideoDecoderCache* VideoDecoderCache::GetInstance() {
  static base::NoDestructor<VideoDecoderCache> instance;
  return instance.get();
}

VideoDecoderCache::VideoDecoderCache() = default;

VideoDecoderCache::~VideoDecoderCache() {
  // VideoDecoderCache is a singleton, so it shouldn't be destroyed.
  // But if it is, the unique_ptrs in cache_ will be destroyed, which destroys
  // the MediaDecoders.
}

bool VideoDecoderCache::CacheKey::operator==(const CacheKey& other) const {
  return codec == other.codec && output_mode == other.output_mode &&
         frame_width == other.frame_width &&
         frame_height == other.frame_height && max_width == other.max_width &&
         max_height == other.max_height && fps == other.fps;
}

std::ostream& operator<<(std::ostream& os,
                         const VideoDecoderCache::CacheKey& key) {
  os << "{codec=" << GetMediaVideoCodecName(key.codec)
     << ", output_mode=" << GetPlayerOutputModeName(key.output_mode)
     << ", frame_width=" << key.frame_width
     << ", frame_height=" << key.frame_height << ", max_width=";
  if (key.max_width) {
    os << *key.max_width;
  } else {
    os << "nullopt";
  }
  os << ", max_height=";
  if (key.max_height) {
    os << *key.max_height;
  } else {
    os << "nullopt";
  }
  os << ", fps=" << key.fps << "}";
  return os;
}

std::string VideoDecoderCache::Put(const CacheKey& key,
                                   std::unique_ptr<MediaDecoder> decoder) {
  if (!decoder) {
    return "error:Invalid: decoder is null";
  }

  if (decoder->is_secure()) {
    return "policy: we don't cache secure decoders";
  }

  // We caches decoder only for short.
  const bool is_cacheable_size = [&] {
    if (key.frame_width == 1080 && key.frame_height == 1920) {
      return true;
    }
    if (key.frame_width == 720 && key.frame_height == 1280) {
      return true;
    }
    return false;
  }();
  if (!is_cacheable_size) {
    return "policy: we don't cache the given resolution";
  }

  // With the new stop() approach, the decoder is already stopped by
  // MediaDecoder::Suspend() (which calls media_codec_bridge->Stop()).
  // We just need to store it.

  std::lock_guard lock(mutex_);
  SB_LOG(INFO) << "Caching video decoder for key=" << key
               << ", num_cached=" << cache_.size();

  if (cache_.size() >= kMaxCacheSize) {
    cache_.pop_front();
  }
  cache_.push_back({key, std::move(decoder)});
  return "";
}

std::unique_ptr<MediaDecoder> VideoDecoderCache::Get(const CacheKey& key) {
  std::lock_guard lock(mutex_);
  for (auto it = cache_.begin(); it != cache_.end(); ++it) {
    if (it->key == key) {
      SB_LOG(INFO) << "Getting cached video decoder for key=" << key;
      std::unique_ptr<MediaDecoder> decoder = std::move(it->decoder);
      cache_.erase(it);
      return decoder;
    }
  }
  return nullptr;
}

}  // namespace starboard::android::shared
